mqtt/mqtt-bridge/src/client.rs

```diff
@@ -184,6 +186,36 @@ impl BridgeIoSource {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct InFlightPublishHandle {
+    publish_handle: PublishHandle,
+    permits: Arc<Semaphore>,
+}
+
+impl InFlightPublishHandle {
+    fn new(publish_handle: PublishHandle, permits: Arc<Semaphore>) -> Self {
+        Self {
+            publish_handle,
+            permits,
+        }
+    }
+
+    pub async fn publish(&self, publication: Publication) {
+        let mut publish_handle = self.publish_handle.clone();
+        let permits = self.permits.clone();
+
+        let fut = async move {
+            let permit = permits.acquire().await;
+            if let Err(e) = publish_handle.publish(publication).await {
+                error!(message = "failed to publish", err = %e);
+            }
+            drop(permit)
+        };
+
+        tokio::spawn(fut);
```

Comment:
why do we need spawn here?

Reply:
We need to spawn so that a separate task can wait for the permit and the original call to publish can complete. That way the caller can keep calling publish without having to manage all the futures. 

For the in-flight draining though, I think what we want to do is return a join handle from this future so we can confirm all the in-flights are complete at time of shutdown.